### ISSUE TITLE
RFC: KQL vs GraphQL

### ISSUE THREAD

**Issue by meyer** [getkirby/kql#1]
I’ve been casually following the development this plugin and I think it’s pretty neat. I like seeing innovation in this area. I’m creating this issue to hopefully prompt some discussion about adopting an existing standard vs creating a new one.

---

GraphQL has one major advantage over KQL: it’s a standard that’s been adopted all over the place. This has resulted in a battle-tested design that’s got a ton of tooling built around it.

### Design

The GraphQL spec describes a ton of functionality but I’ve highlighted a few of my favourite features:
- Predictability: Data is always a known shape.
- Nullability: Developers can reliably know whether or not a field is potentially null, which leads to much more resiliently-built websites.
- Standard error handling: Resolution failures bubble up to the nearest nullable field. If something in a field resolver goes catastrophically wrong, the damage is contained in a standard way.
- Partial responses/response with errors: If one or more fields fail to resolve, a partial response is sent with an accompanying error array that shows the fields failed to resolve and the specific error per-field. Errors can be handled differently on a per-request basis.

### Tooling

The GraphQL community has built all sorts of nifty things around the GraphQL standard. There are IDE/text editor integrations that provide autocomplete as you’re writing queries. Queries are syntax highlighted just about everywhere. There are GraphQL server and client implementations for just about every web-friendly language.

One of my recent favourite community-built tools is [`graphql-codegen`](https://graphql-code-generator.com/). Here’s how I use it in my personal projects:

- I generate TypeScript types for API calls. GraphQL types and enums translate directly to TypeScript interfaces and enums. Nullability transfers over in the types so when I’m digging through a query response object, I know whether or not I’m accessing something safely.

- I generate async fetch functions for each GraphQL query:

  This query:
  ```gql
  query exampleQuery($limit: Int = 10, $offset: Int = 0) {
    page(id: "photography") {
      children(limit: $limit, offset: $offset) {
        title
        text
        images {
          url
          srcset
        }
      }
    }
  }
  ```

  becomes this fetch function:

  ```ts
  const exampleQueryText = `...`;

  export const exampleQuery = (params: ExampleQueryParams, options?: ApolloQueryOptions): ExampleQueryResult => {
    return apolloClient.query<ExampleQueryParams, ExampleQueryResult>(exampleQueryText, params, options);
  }
  ```

---

Sorry for this lengthy issue. I don’t want to discourage you in the further development of this plugin, but I did want to provide some perspective from the GraphQL side of things. Let me know if you have any questions.

I know KQL has the potential to be quite a bit more powerful than GraphQL but that power comes at the expense of being able to take advantage of a wealth of tools that are available today. For the headless Kirby use case I think that’s a pretty big deal.

**Comment by lukasbestle** [comment]
Thanks for your feedback!

We have experimented with GraphQL during the development of Kirby 3 (while Kirby 3 was in beta) and were quite happy about the possibilities and available tools as well. However the standard turned out to be not a good fit for Kirby and we therefore had to switch back to a REST API (after a lot was already implemented using GraphQL actually).

The most important reasons it wasn't a good fit are:

- The advantage of predictability you are describing makes the implementation for a flexible CMS like Kirby a nightmare: Kirby objects (pages, files, users) don't have a fixed "shape", but a set of fields that can be freely configured by the user. To bring that in line with the GraphQL frontend, the schema would need to be generated on the fly based on the specific data of the site – which makes it all less predictable again and also increases complexity by a lot: You really wouldn't want to have potentially dozens or hundreds of types in the schema.
- The query syntax is weirdly JSON-like, but not actually JSON. This means that tooling is actually not only *available*, but also basically *required* to use GraphQL properly (if you don't like spaghetti code). You can't just quickly write a query and pass that to the HTTP request library/tool of your choice. Kirby is known for being easy to use and easy to get started with while at same time offering many advanced features you can add to the site later in the process of site development. GraphQL on the other hand can feel pretty advanced for beginners even when only simple queries are used: Without a library, the query needs to be written entirely by hand, native language features like `JSON.stringify()` can't be used.

**Comment by meyer** [comment]
Ah yeah, I remember discussing GraphQL stuff briefly with Bastian in Slack way back in the day. I had halted development on a GraphQL plugin I was working on when I heard there might be first-party GraphQL support in K3, but he encouraged me to continue working on it when things went back to REST.

- Re: GraphQL tooling being required, that’s only partially true—GET support is turned off by default in most GraphQL server applications, but mainly to prevent PII from appearing in the URL and potentially being logged (for this very reason, somebody suggested moving KQL from GET to POST in the Slack). In my personal projects I enable GET support in development. Test URLs end up looking like so:

  ```
  https://localhost:1234/graphql?query="query getPost($id: Int) { post(id: $id) { title slug date author } }"&variables='{"id": 1234}'
  ```

  Surprisingly readable—the worst part of that query (imo) is the hand-written JSON.

  I do mostly use GraphQL query browsers though, and mainly for convenience. GraphiQL has a query history that I can reference, and it autocompletes my queries and shows docstring comments inline, which really helps with API discoverability:
  
  ![Screenshot 2020-01-21 14 37 44](https://user-images.githubusercontent.com/13781/72856839-a3c7b780-3c70-11ea-84c4-7ce84bb34311.png)

  Network requests are pretty easy to debug as well. Not much to them:

  ![Screenshot 2020-01-21 15 32 44](https://user-images.githubusercontent.com/13781/72856877-bf32c280-3c70-11ea-86d2-a0290f3c541b.png)


- Re: syntax being JSON-y but not JSON, I know what you mean. A lot of folks get hung up on this; I did too for a long time. It’s pretty weird to begin with, but you get used to it, and it starts to make sense after a while. It looks like JSON for a reason—when you write queries, you’re writing a description of _hierarchical data_. The brackets make it look JSON-y but they’re only present because whitespace sensitivity in a query language would be a bit hair-raising.

  Someone has already suggested allowing variable placeholders—a feature already present in GraphQL. At a certain point, the fact that the chosen query syntax is JSON starts to matter less. What’s the advantage of JSON? I suppose you could construct deeply nested queries programmatically or conditionally add fields, but the choice of JSON seems like an implementation detail since it makes _parsing_ the query easier. I’m going to be _writing_ queries though, so that advantage is lost on me.

  From my perspective, KQL syntax is a stone’s throw away from being GraphQL, just a liiiiittle bit different.

---

With the advent of KQL I figured now is the time to advocate for GraphQL. If there was a way to describe what data looked like, everything else would fairly naturally fall into place.

For an API with potentially harmful side effects—for example, a thumbnailing query that doesn’t rate limit or restrict dimensions—it might make sense to require users to choose what to expose. If that sort of thing becomes opt-in, doing something like dynamically generating a GraphQL schema would be pretty straightforward.

If I get some free time this weekend I’ll try cook up a proof of concept. It’s probably easier to show with code than to try to describe what I’m saying.

**Comment by lukasbestle** [comment]
> Re: GraphQL tooling being required, that’s only partially true—GET support is turned off by default

I think the issue is not that tools are required/encouraged during development, but also in production. As I said, you can't use `JSON.stringify()` on structured data and call it a day. You need to construct the query as a string in the GraphQL query syntax. You need abstractions to make this easier, which makes everything a lot more complex for beginners.

Bastian has posted examples of KQL requests written in vanilla JS in Slack that looked very very clean. That's not as easy to do with GraphQL.

> If I get some free time this weekend I’ll try cook up a proof of concept. It’s probably easier to show with code than to try to describe what I’m saying.

That would be great! I think it's always great to have choice and alternatives.

As I wrote, the issue will probably be creating a flexible schema that supports the different fields of each blueprint without creating dozens of different types. If you know a solution to that, I'm really looking forward to it.

**Comment by bastianallgeier** [comment]
I totally get all your points and I totally agree that there are many benefits to follow a standard like GraphQL. Especially the tooling part is a huge plus! 

But I've already tried my very best to get good GraphQL support working for Kirby and I failed because of the missing content structure. I was not able to find a way to generate schemas based on our blueprints on the fly. I have to say that this has been a while ago and maybe there are now ways to solve this, but back then it was a massive pain in the ass to build a good GraphQL server with PHP. 

In the end I was only able to add content as an array of key, value objects to the schema and that felt terrible. 

I think GraphQL could still coexist with KQL. I see KQL more as a replacement for what we currently have in our REST API. The REST API has quite a lot of limitations right now, which doesn't make it that attractive at all. We experience this ourselves in the Panel where we often have to add additional custom API endpoints for fields or sections because the REST API doesn't cut it. 

One of the most important parts of KQL is its onion layer approach to security. This was a huge task, but now we have a setup that only allows methods in the API, which can be executed without risk. 

This architecture can be reused for a GraphQL implementation in my opinion and already gives us a big head start. But the missing piece is the dynamic schema. If there's anyone out there to help with this, I'm more than happy to work on a GraphQL plugin together.
